### PR TITLE
[LLVM][DWARF] Chnage order for .debug_names abbrev print out

### PR DESCRIPTION
**Pull request by ayermolo**
This stemps from conversatin in: https://github.com/llvm/llvm-project/pull/77457#discussion_r1457889792.
Right now Abbrev code for abbrev is combination of DIE TAG and other attributes.
In the future it will be changed to be an index. Since DenseSet does not
preserve an order, added a sort based on abbrev code. Once change to index is
made, it will print out abbrevs in the order they are stored.

### CODE REVIEW

**Comment by ayermolo** [comment]
Alternative is to keep a vector as abbrevs read in. I looked into SetVector but doesn't look like it has find api. Used in NameIndex::getEntry. Also seems a tad wasteful since it only matters for printing out.

**Review by clayborg — COMMENTED** [review]
When we dump things we should dump them in the order in which they appear in the `.debug_xxx` section. Here we are sorting and dumping them in some order. I realize eventually the abbrevs will have an index and that will make the order more natural, but when dumping the contents of the section, we should do it in the order in which they are defined in the file at all times since we are wanting to see the contents of the section we asked to dump. There aren't usually that many abbreviation combinations right? So maybe the dump function can parse the info again and dump each entry it finds in the right order and throw away the temp `llvm::DWARFDebugNames::Abbrev` item as it dumps each one? Or we can maintain an offset in each `llvm::DWARFDebugNames::Abbrev` entry and then sort by that when dumping?

**Comment by ayermolo** [comment]
> When we dump things we should dump them in the order in which they appear in the `.debug_xxx` section. Here we are sorting and dumping them in some order. I realize eventually the abbrevs will have an index and that will make the order more natural, but when dumping the contents of the section, we should do it in the order in which they are defined in the file at all times since we are wanting to see the contents of the section we asked to dump. There aren't usually that many abbreviation combinations right? So maybe the dump function can parse the info again and dump each entry it finds in the right order and throw away the temp `llvm::DWARFDebugNames::Abbrev` item as it dumps each one? Or we can maintain an offset in each `llvm::DWARFDebugNames::Abbrev` entry and then sort by that when dumping?

I agree we should print in the same order as it's in section.
Changed to sort by offset.

File: llvm/include/llvm/DebugInfo/DWARF/DWARFAcceleratorTable.h [hunk 1]
```diff
@@ -413,12 +413,14 @@ class DWARFDebugNames : public DWARFAcceleratorTable {
   /// Abbreviation describing the encoding of Name Index entries.
   struct Abbrev {
     uint32_t Code;  ///< Abbreviation code
+    uint64_t AbbrevOffset; /// < Abbreviation offset in the .debug_names section
```
Comment (clayborg) [inline llvm/include/llvm/DebugInfo/DWARF/DWARFAcceleratorTable.h:416]: For padding reasons, I would move this to be the first item in this struct.

File: llvm/lib/DebugInfo/DWARF/DWARFAcceleratorTable.cpp [hunk 2]
```diff
@@ -540,7 +540,7 @@ DWARFDebugNames::NameIndex::extractAbbrev(uint64_t *Offset) {
     return createStringError(errc::illegal_byte_sequence,
                              "Incorrectly terminated abbreviation table.");
   }
-
+  uint64_t AbbrevOffset = *Offset;
```
Comment (clayborg) [inline llvm/lib/DebugInfo/DWARF/DWARFAcceleratorTable.cpp:543]: make const to ensure someone doesn't change it?

**Review by clayborg — COMMENTED** [review]
A few quick things in the inline comments and this will be good to go.

**Comment by dwblaikie** [comment]
How's this compare to how we handle .debug_abbrevs? (perhaps we could be sharing some parsing infrastructure, the same as I'm suggesting/hoping we share some generation infrastructure - but even if not shared code, bringing two different implementations into alignment so they do/express things more similarly would be good)

File: llvm/lib/DebugInfo/DWARF/DWARFAcceleratorTable.cpp [hunk 3]
```diff
@@ -847,8 +847,15 @@ void DWARFDebugNames::NameIndex::dumpForeignTUs(ScopedPrinter &W) const {
 
 void DWARFDebugNames::NameIndex::dumpAbbreviations(ScopedPrinter &W) const {
   ListScope AbbrevsScope(W, "Abbreviations");
-  for (const auto &Abbr : Abbrevs)
-    Abbr.dump(W);
+  std::vector<const Abbrev *> AbbrevsVect;
+  for (const llvm::DWARFDebugNames::Abbrev &Abbr : Abbrevs)
```
Comment (felipepiovezan) [inline llvm/lib/DebugInfo/DWARF/DWARFAcceleratorTable.cpp:851]: we're inside the `llvm` namespace already

File: llvm/lib/DebugInfo/DWARF/DWARFAcceleratorTable.cpp [hunk 4]
```diff
@@ -847,8 +847,15 @@ void DWARFDebugNames::NameIndex::dumpForeignTUs(ScopedPrinter &W) const {
...
+    AbbrevsVect.push_back(&Abbr);
+  std::sort(AbbrevsVect.begin(), AbbrevsVect.end(),
+            [](const Abbrev *LHS, const Abbrev *RHS) {
+              return LHS->AbbrevOffset < RHS->AbbrevOffset;
+            });
+  for (const llvm::DWARFDebugNames::Abbrev *Abbr : AbbrevsVect)
```
Comment (felipepiovezan) [inline llvm/lib/DebugInfo/DWARF/DWARFAcceleratorTable.cpp:857]: no need for `llvm::`

**Comment by ayermolo** [comment]
> How's this compare to how we handle .debug_abbrevs? (perhaps we could be sharing some parsing infrastructure, the same as I'm suggesting/hoping we share some generation infrastructure - but even if not shared code, bringing two different implementations into alignment so they do/express things more similarly would be good)

The DeclSet are stored in the map based on offset. Within it's a Decls vector. Looks like lookups in Decls vector is linear (iterates over it). At least in DWARFAbbreviationDeclarationSet.

Not sure how this can be replicated generically (before we switch to index) with just std::map for .debug_names abbrev. Because we need to lookup based on AbbrevTag (DWARFDebugNames::NameIndex::getEntry), but keep it sorted by offset (for llvm-dwarfdump).

Maybe we can re-visit after llvm is switched to output indices? I am finishing BOLT implementation and adding tests. So would help if llvm-dwarfdump outputs things in correct order. :)

**Comment by clayborg** [comment]
> How's this compare to how we handle .debug_abbrevs? (perhaps we could be sharing some parsing infrastructure, the same as I'm suggesting/hoping we share some generation infrastructure - but even if not shared code, bringing two different implementations into alignment so they do/express things more similarly would be good)

The abbrevs currently makes a DWARFAbbreviationDeclarationSet which contains a `std::vector<DWARFAbbreviationDeclaration> Decls;` and it also maintains a value call `FirstAbbrCode` which is set to UINT32_MAX if the abbrevs were not indexed starting with some number. Compilers like to use 1 and the starting number. Then when we ask for an abbrev from the DWARFAbbreviationDeclarationSet, it sees is `FirstAbbrCode` is not set to UINT32_MAX, and if it isn't it returns a direct access using:
```
return &Decls[AbbrCode - FirstAbbrCode];
```
So O(1) lookups. Else it falls back to a very costly linear search. But most compilers emit the abbrevs with and index so this works well for 99% of the cases.

The DWARFDebugNames::Entry is quite different and stored in a hash map as the abbrev codes are not indexed.

File: llvm/lib/DebugInfo/DWARF/DWARFAcceleratorTable.cpp [hunk 5]
```diff
@@ -847,8 +847,15 @@ void DWARFDebugNames::NameIndex::dumpForeignTUs(ScopedPrinter &W) const {
 
 void DWARFDebugNames::NameIndex::dumpAbbreviations(ScopedPrinter &W) const {
   ListScope AbbrevsScope(W, "Abbreviations");
-  for (const auto &Abbr : Abbrevs)
-    Abbr.dump(W);
+  std::vector<const Abbrev *> AbbrevsVect;
+  for (const llvm::DWARFDebugNames::Abbrev &Abbr : Abbrevs)
+    AbbrevsVect.push_back(&Abbr);
+  std::sort(AbbrevsVect.begin(), AbbrevsVect.end(),
```
Comment (felipepiovezan) [inline llvm/lib/DebugInfo/DWARF/DWARFAcceleratorTable.cpp:853]: We should never use `std::sort`, instead using the range-based (with some shuffling magic for expensive asserts) sort from stl extras:

```
sort(AbbrevVect, [] (...){});
```

https://llvm.org/docs/CodingStandards.html#beware-of-non-deterministic-sorting-order-of-equal-elements

**Review by felipepiovezan — APPROVED** [review]
LGTM!

File: llvm/lib/DebugInfo/DWARF/DWARFAcceleratorTable.cpp (hunk 5 above)
Comment (ayermolo) [reply llvm/lib/DebugInfo/DWARF/DWARFAcceleratorTable.cpp:853]: My interpretation that only applies if we have equal elements. Which shouldn't be the case in this case. Abbrev offsets are monotonically increasing.

**Comment by ayermolo** [comment]
Can we revisit this after llvm switches to indices?

**Comment by clayborg** [comment]
> Can we revisit this after llvm switches to indices?

I am fine with taking care of switching to a more DebugAbbrev like solution later.

File: llvm/lib/DebugInfo/DWARF/DWARFAcceleratorTable.cpp [hunk 6]
```diff
@@ -847,8 +847,15 @@ void DWARFDebugNames::NameIndex::dumpForeignTUs(ScopedPrinter &W) const {
 
 void DWARFDebugNames::NameIndex::dumpAbbreviations(ScopedPrinter &W) const {
   ListScope AbbrevsScope(W, "Abbreviations");
-  for (const auto &Abbr : Abbrevs)
-    Abbr.dump(W);
+  std::vector<const Abbrev *> AbbrevsVect;
+  for (const DWARFDebugNames::Abbrev &Abbr : Abbrevs)
+    AbbrevsVect.push_back(&Abbr);
+  llvm::sort(AbbrevsVect.begin(), AbbrevsVect.end(),
```
Comment (felipepiovezan) [inline llvm/lib/DebugInfo/DWARF/DWARFAcceleratorTable.cpp:853]: minor nit: note that we can use a range based version here, no need for `begin` and `end`, just `AbbrevsVect` :)

**Comment by dwblaikie** [comment]
> > How's this compare to how we handle .debug_abbrevs? (perhaps we could be sharing some parsing infrastructure, the same as I'm suggesting/hoping we share some generation infrastructure - but even if not shared code, bringing two different implementations into alignment so they do/express things more similarly would be good)
> 
> The abbrevs currently makes a DWARFAbbreviationDeclarationSet which contains a `std::vector<DWARFAbbreviationDeclaration> Decls;` and it also maintains a value call `FirstAbbrCode` which is set to UINT32_MAX if the abbrevs were not indexed starting with some number. Compilers like to use 1 and the starting number. Then when we ask for an abbrev from the DWARFAbbreviationDeclarationSet, it sees is `FirstAbbrCode` is not set to UINT32_MAX, and if it isn't it returns a direct access using:
> 
> ```
> return &Decls[AbbrCode - FirstAbbrCode];
> ```
> 
> So O(1) lookups. Else it falls back to a very costly linear search. But most compilers emit the abbrevs with and index so this works well for 99% of the cases.
> 
> The DWARFDebugNames::Entry is quite different and stored in a hash map as the abbrev codes are not indexed.

I'm not sure I understand - what do you mean by "the abbrev codes are not indexed"? Because of LLVM's current output that uses the weird bit fiddling?

I'd consider that a bug/suboptimality, and I'd be fine with llvm-dwarfdump devolving to a linear search through abbrevs in the case where the abbrevs are not monotonically increasing.

**Comment by clayborg** [comment]
> > So O(1) lookups. Else it falls back to a very costly linear search. But most compilers emit the abbrevs with and index so this works well for 99% of the cases.
> > The DWARFDebugNames::Entry is quite different and stored in a hash map as the abbrev codes are not indexed.
> 
> I'm not sure I understand - what do you mean by "the abbrev codes are not indexed"? Because of LLVM's current output that uses the weird bit fiddling?

The abbreviation codes are not 1 based indexes like they are for .debug_abbrev:
```
  Abbreviations [
    Abbreviation 0x1718 {
      Tag: DW_TAG_subprogram
      DW_IDX_die_offset: DW_FORM_ref4
      DW_IDX_parent: DW_FORM_flag_present
    }
    Abbreviation 0xb18 {
      Tag: DW_TAG_typedef
      DW_IDX_die_offset: DW_FORM_ref4
      DW_IDX_parent: DW_FORM_flag_present
    }
    Abbreviation 0x998 {
      Tag: DW_TAG_structure_type
      DW_IDX_die_offset: DW_FORM_ref4
      DW_IDX_parent: DW_FORM_flag_present
    }
    Abbreviation 0x1218 {
      Tag: DW_TAG_base_type
      DW_IDX_die_offset: DW_FORM_ref4
      DW_IDX_parent: DW_FORM_flag_present
    }
  ]
```

> 
> I'd consider that a bug/suboptimality, and I'd be fine with llvm-dwarfdump devolving to a linear search through abbrevs in the case where the abbrevs are not monotonically increasing.

I was mostly commenting that .debug_abbrev code lookup is O(1) where if we don't use indexes for the abbrev codes we must do some sort of search and if this happens in .debug_abbrev the search is linear. And of course any code that tries to find abbrevs should work (linear or direct access) depending on what the input is. The main point is it is much more efficient to use 1 based indexes since zero is reserved so that we _can_ do effecient lookups. Most of what I was pointing out was around what .debug_abbrev does right and how it is different from what we are doing for .debug_names abbrev codes.

**Comment by dwblaikie** [comment]
> I was mostly commenting that .debug_abbrev code lookup is O(1) where if we don't use indexes for the abbrev codes we must do some sort of search and if this happens in .debug_abbrev the search is linear. And of course any code that tries to find abbrevs should work (linear or direct access) depending on what the input is. The main point is it is much more efficient to use 1 based indexes since zero is reserved so that we _can_ do effecient lookups. Most of what I was pointing out was around what .debug_abbrev does right and how it is different from what we are doing for .debug_names abbrev codes.

Fair - thanks for helping me understand what we're doing today (on the LLVM emission side - we aren't emitting monotonically increasing debug_names abbrevation numbers, compared to debug_abbrev where we do use monotonically increasing abbrev numbers).

I think that's a mistake in LLVM's emission code, and not one we should worry about when designing llvm-dwarfdump. Both sides (LLVM DWARF emission, and llvm-dwarfdump parsing) should be fixed (to behave similarly to - but I don't think we need to wait for the emission to get better/fixed before we make llvm-dwarfdump handle things well)

I'd prefer to see the llvm-dwarfdump code not use a hash map, ideally reuse a generalized form of the debug_abbrev handling code - that's efficient for monotonically increasing abbrev numbers, and falls back to a linear search otherwise - and the printing should print from that list that's stored in the same order it's read from the input - and that list can be directly indexed if it's monotonically increasing, or linearly searched if it's not.

**Comment by ayermolo** [comment]
OK, put up a PR for .debug_names in BOLT (no parent index support for now). Let me circle back to this, and change implementation to be sequential on LLVM side.

**Pull request merged**